src/main/java/software/amazon/ionpathextraction/SearchPath.java

```diff
@@ -13,21 +13,64 @@
 
 package software.amazon.ionpathextraction;
 
+import java.util.List;
 import java.util.function.BiFunction;
 import software.amazon.ion.IonReader;
+import software.amazon.ionpathextraction.internal.ArrayUtils;
+import software.amazon.ionpathextraction.internal.MatchContext;
+import software.amazon.ionpathextraction.pathcomponents.PathComponent;
 
 /**
  * A path which is provided to the extractor for matching.
  *
  * @param <T> type accepted by the callback function
  */
-interface SearchPath<T> {
+final class SearchPath<T> {
 
-    BiFunction<IonReader, T, Integer> getCallback();
+    private final List<PathComponent> pathComponents;
+    private final BiFunction<IonReader, T, Integer> callback;
+    private final String[] annotations;
 
-    Type getType();
+    SearchPath(final List<PathComponent> pathComponents,
+               final BiFunction<IonReader, T, Integer> callback,
+               final String[] annotations) {
+        this.annotations = annotations;
+        this.pathComponents = pathComponents;
+        this.callback = callback;
+    }
+
+    int size() {
+        return pathComponents.size();
+    }
+
+    boolean isTerminal(final int pathComponentIndex) {
```

Comment:
I realize this is an internal class, but it would be helpful to have a comment explaining the purpose of each method.

Reply:
You are right I got lazy, will add javadoc to this. 

Will try to link the implementors guide as well since most of the terms are defined there